src/records.rs

```diff
@@ -0,0 +1,303 @@
+    }
+
+    pub fn nullifier(&self) -> &Nullifier {
+        &self.nullifier
+    }
+
+    pub fn on_hold(&self, now: u64) -> bool {
+        matches!(self.hold_until, Some(t) if t > now)
+    }
+}
+
+/// An editor to update a record or records store
+pub struct RecordEditor<'a> {
+    record: Record,
+    store: &'a mut Records,
+}
+impl<'a> RecordEditor<'a> {
+    /// Create a record editor.
+    fn new(store: &'a mut Records, record: Record) -> Self {
+        Self { record, store }
+    }
+
+    /// Update when this record will be on hold until if it is pending
+    /// Set this when the record is part of a transaction that isn't confirmed
+    pub fn hold_until(mut self, until: u64) -> Self {
+        self.record.hold_until = Some(until);
+        self
+    }
+
+    /// Update the record to no longer be on hold
+    pub fn unhold(mut self) -> Self {
+        self.record.hold_until = None;
+        self
+    }
+
+    /// Save the record to the store.
+    ///
+    /// Returns the stored record.
+    pub fn save<L: Ledger>(&mut self) -> Result<Record, KeystoreError<L>> {
+        self.store.store(self.record.uid, &self.record)?;
+        Ok(self.record.clone())
+    }
+}
+
+impl<'a> Deref for RecordEditor<'a> {
+    type Target = Record;
+
+    fn deref(&self) -> &Record {
+        &self.record
+    }
+}
+
+impl<'a> DerefMut for RecordEditor<'a> {
+    fn deref_mut(&mut self) -> &mut Record {
+        &mut self.record
+    }
+}
+
+type RecordsStore = KeyValueStore<u64, Record>;
+
+pub struct Records {
+    store: RecordsStore,
+    /// Record (size, uid) indexed by asset type, owner, and freeze status, for easy allocation as
+    /// transfer or freeze inputs. The records for each asset are ordered by increasing size, which
+    /// makes it easy to implement a worst-fit allocator that minimizes fragmentation.
+    asset_records:
+        PersistableHashMapBTreeMultiMap<(AssetCode, UserAddress, FreezeFlag), (RecordAmount, u64)>,
+    /// Record uids indexed by nullifier, for easy removal when confirmed as transfer inputs
+    nullifier_records: PersistableHashMap<Nullifier, u64>,
+}
+
+impl Records {
+    #![allow(dead_code)]
+
+    pub fn new<L: Ledger>(
+        loader: &mut AtomicStoreLoader,
+        adaptor: EncryptingResourceAdapter<(u64, Option<Record>)>,
+        fill_size: u64,
+    ) -> Result<Self, KeystoreError<L>> {
+        let log = AppendLog::load(loader, adaptor, "keystore_records", fill_size)?;
+        let store = RecordsStore::new(log)?;
+        let mut records = Self {
+            store,
+            asset_records: Persistable::new(),
+            nullifier_records: Persistable::new(),
+        };
+        records.reload();
+        Ok(records)
+    }
+
+    /// Create a Record
+    ///
+    /// Returns an editor to the newly created Record
+    pub fn create<L: Ledger>(
+        &mut self,
+        uid: u64,
+        ro: RecordOpening,
+        nullifier: Nullifier,
+        hold_until: Option<u64>,
```

Comment:
Since this is optional, it would be cleaner to remove it as a parameter. You can always do `records.create(uid, ro, nullifier)?.hold_until(t).save()`.